source/common/router/router.cc

```diff
@@ -694,6 +704,46 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
       *this, std::move(generic_conn_pool), can_send_early_data, can_use_http3);
   LinkedList::moveIntoList(std::move(upstream_request), upstream_requests_);
   upstream_requests_.front()->acceptHeadersFromRouter(end_stream);
+  if (streaming_shadows_) {
+    // start the shadow streams.
+    for (const auto& shadow_policy_wrapper : active_shadow_policies_) {
+      const auto& shadow_policy = shadow_policy_wrapper.get();
+      const absl::optional<absl::string_view> shadow_cluster_name =
+          getShadowCluster(shadow_policy, *downstream_headers_);
+      if (!shadow_cluster_name.has_value()) {
+        continue;
+      }
+      auto shadow_headers = Http::createHeaderMap<Http::RequestHeaderMapImpl>(*shadow_headers_);
+      auto options = Http::AsyncClient::RequestOptions()
+                         .setTimeout(timeout_.global_timeout_)
+                         .setParentSpan(callbacks_->activeSpan())
+                         .setChildSpanName("mirror")
+                         .setSampled(shadow_policy.traceSampled())
+                         .setIsShadow(true)
+                         .setBufferAccount(callbacks_->account())
+                         .setBufferLimit(callbacks_->decoderBufferLimit());
+      if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.closer_shadow_behavior")) {
+        options.setFilterConfig(config_);
+      }
+      if (end_stream) {
+        // This is a header-only request, and can be dispatched immediately to the shadow
+        // without waiting.
+        Http::RequestMessagePtr request(new Http::RequestMessageImpl(
+            Http::createHeaderMap<Http::RequestHeaderMapImpl>(*shadow_headers_)));
+        config_.shadowWriter().shadow(std::string(shadow_cluster_name.value()), std::move(request),
+                                      options);
+      } else {
+        Http::AsyncClient::OngoingRequest* shadow_stream = config_.shadowWriter().streamingShadow(
+            std::string(shadow_cluster_name.value()), std::move(shadow_headers), options);
+        if (shadow_stream != nullptr) {
+          shadow_streams_[shadow_stream] = end_stream;
+          shadow_stream->setDestructorCallback(
+              [this, shadow_stream]() { shadow_streams_.erase(shadow_stream); });
```

Comment:
what's the guarantee we have that we won't call this dtor under iterating this map? Can we have this as an assertion, or something programmatic?